### PR TITLE
Vote System: Approval Voting

### PR DESCRIPTION
**Pull request by lessthnthree**
## About The Pull Request

Approval Voting is a system in which voters can select as many maps as they want, instead of selecting only one. Final tallies show how many votes each map received, and the winner is the map with the most support.

## Changes since https://github.com/tgstation/tgstation/pull/73413
- Custom votes can now be started using either system
- Icon during AV votes indicating your selections
- Map population filter counts active players and participating ghosts

https://user-images.githubusercontent.com/83487515/222580901-61506cc3-dc42-4435-9775-1e6291a3f734.mp4

## Why It's Good For The Game

First-past-the-post (our current voting system) has flaws such as creating a bunch of wasted votes, in that a large number of selections ultimately have no impact and for example, a map can win a 3 way race 11/10/10, even though 2/3 of the votes were not for that map. This leads to people having to vote strategically, and perhaps not what their true choice is.

Approval Voting solves this by instead allowing the player to select all the maps they would like to play, so they can vote for their true preferred choice, as well as alternates.

For example, a player that wants Metastation, is okay with Icebox, and doesn't want Delta may feel pressured to vote Icebox if it's in a 2 way race with Delta.

AV lets them vote for Meta, and Icebox or as many others as they want as their alternates and creates a more fair outcome of a map vote.

Map population filter removing AFK/lobby screen dwellers gives a better number of active players so as to not trip the map filter's population cap earlier than it should.

tl;dr: Less of this

image

## Changelog

:cl: LT3
rscadd: Added new multi-vote system
balance: Map votes are now calculated using multi-vote instead of the old single-vote system
admin: Admins can now use either multi-vote or single-vote for custom votes
code: Map choice filtering uses active player count, not all connected clients
/:cl:

### CODE REVIEW

**Comment by Iamgoofball** [comment]
👎 this is a massive nerf to tram

**Comment by Mothblocks** [comment]
That commit you just pushed made the code worse, procs shouldn't do radically different things like that

File: code/__DEFINES/subsystems.dm [hunk 1]
```diff
@@ -331,3 +331,7 @@
 
 /// The timer key used to know how long subsystem initialization takes
 #define SS_INIT_TIMER_KEY "ss_init"
+
+// Vote subsystem counting methods
+#define VOTE_COUNT_METHOD_FPTP 1
+#define VOTE_COUNT_METHOD_AV 2
```
Comment (Mothblocks) [inline code/__DEFINES/subsystems.dm:337]: You need to document what these are so people don't have to google FPTP and AV

Comment (lessthnthree) [reply code/__DEFINES/subsystems.dm]: Ah yeah, got it.

File: tgui/packages/tgui/interfaces/VotePanel.tsx [hunk 2]
```diff
@@ -26,12 +26,15 @@ type ActiveVote = {
   question: string | null;
   timeRemaining: number;
   choices: Option[];
+  countMethod: number;
```
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/VotePanel.tsx:29]: Because this is TypeScript, use an enum with `= 1` and `= 2` so that we can actually read the values of this

File: code/controllers/subsystem/vote.dm [hunk 3]
```diff
@@ -111,6 +115,32 @@ SUBSYSTEM_DEF(vote)
...
+/**
+ * Submit a vote using the Approval Voting (AV) system
+ * Any number of selections per person, and the selection with the most votes wins.
+ */
+/datum/controller/subsystem/vote/proc/submit_av_vote(mob/voter, their_vote)
+	if(!current_vote)
+		return
+	if(!voter?.ckey)
+		return
+	if(CONFIG_GET(flag/no_dead_vote) && voter.stat == DEAD && !voter.client?.holder)
```
Comment (Mothblocks) [inline code/controllers/subsystem/vote.dm:130]: Was this the code that was here in the other votes? The condition here confuses me a little

File: code/datums/votes/custom_vote.dm [hunk 4]
```diff
@@ -1,9 +1,14 @@
 /// The max amount of options someone can have in a custom vote.
 #define MAX_CUSTOM_VOTE_OPTIONS 10
 
-/datum/vote/custom_vote
-	name = "Custom"
-	message = "Click here to start a custom vote."
+/datum/vote/custom_vote/fptp
+	name = "Custom FPTP"
+	message = "Click here to start a custom FPTP vote, FPTP (one selection per voter)"
```
Comment (Mothblocks) [inline code/datums/votes/custom_vote.dm:6]: I think we should probably keep this intuitive as "simple vote" / "multi-vote". It's significantly easier for everyone to understand. FPTP and AV are way too jargon-y for me.

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Some smaller stuff

File: code/datums/votes/custom_vote.dm (hunk 4 above)
Comment (lessthnthree) [reply code/datums/votes/custom_vote.dm:6]: okay

File: code/controllers/subsystem/vote.dm (hunk 3 above)
Comment (lessthnthree) [reply code/controllers/subsystem/vote.dm:130]: Yeah that's existing code.

Comment (Mothblocks) [reply code/controllers/subsystem/vote.dm:130]: Alright, sounds fine, I was trying to check in my head if this actually works

File: tgui/packages/tgui/interfaces/VotePanel.tsx (hunk 2 above)
Comment (lessthnthree) [reply tgui/packages/tgui/interfaces/VotePanel.tsx:29]: Alrighty!

**Comment by lessthnthree** [comment]
Instead of FPTP/AV it's now renamed Single/Multi vote
